### PR TITLE
Hook up effect modeling for `TypeVar` and `UnionAll`

### PR DESCRIPTION
**Pull request by Keno**
We already had effect modeling for `_typevar`, which `TypeVar` should just defer to. Add a simple model for `UnionAll` as well, though in the future we can add the Vararg case also.

### CODE REVIEW

File: base/compiler/abstractinterpretation.jl [hunk 1]
```diff
@@ -1834,35 +1834,38 @@ end
...
                 tv = a2.val
             elseif isa(a2, PartialTypeVar)
                 tv = a2.tv
                 canconst = false
             else
-                return Any
+                return CallMeta(Any, Effects(EFFECTS_TOTAL; nothrow=false), NoCallInfo())
             end
-            !isa(tv, TypeVar) && return Any
+            !isa(tv, TypeVar) && return CallMeta(Any, Effects(EFFECTS_TOTAL; nothrow=false), NoCallInfo())
```
Comment (aviatesk) [inline base/compiler/abstractinterpretation.jl:1862]:
```suggestion
            !isa(tv, TypeVar) && return CallMeta(Any, EFFECTS_THROWS, NoCallInfo())
```

File: base/compiler/abstractinterpretation.jl [hunk 2]
```diff
@@ -1834,35 +1834,38 @@ end
...
         if has_free_typevars(body)
-            a2 = argtypes[2]
             if isa(a2, Const)
                 tv = a2.val
             elseif isa(a2, PartialTypeVar)
                 tv = a2.tv
                 canconst = false
             else
-                return Any
+                return CallMeta(Any, Effects(EFFECTS_TOTAL; nothrow=false), NoCallInfo())
```
Comment (aviatesk) [inline base/compiler/abstractinterpretation.jl:1860]:
```suggestion
                return CallMeta(Any, EFFECTS_THROWS, NoCallInfo())
```

File: base/compiler/abstractinterpretation.jl [hunk 3]
```diff
@@ -1834,35 +1834,38 @@ end
...
         elseif isType(a3)
             body = a3.parameters[1]
             canconst = false
         else
-            return Any
+            return CallMeta(Any, Effects(EFFECTS_TOTAL; nothrow), NoCallInfo())
         end
         if !isa(body, Type) && !isa(body, TypeVar)
-            return Any
+            return CallMeta(Any, Effects(EFFECTS_TOTAL; nothrow=false), NoCallInfo())
```
Comment (aviatesk) [inline base/compiler/abstractinterpretation.jl:1851]:
```suggestion
            return CallMeta(Any, EFFECTS_THROWS, NoCallInfo())
```

File: base/compiler/abstractinterpretation.jl [hunk 4]
```diff
@@ -1834,35 +1834,38 @@ end
...
+                a3 ⊑ Type ||
+                a3 ⊑ TypeVar)
         if isa(a3, Const)
             body = a3.val
         elseif isType(a3)
             body = a3.parameters[1]
             canconst = false
         else
-            return Any
+            return CallMeta(Any, Effects(EFFECTS_TOTAL; nothrow), NoCallInfo())
```
Comment (aviatesk) [inline base/compiler/abstractinterpretation.jl:1848]:
```suggestion
            return CallMeta(Any, EFFECTS_THROWS, NoCallInfo())
```

**Review by aviatesk — APPROVED** [review]

File: base/compiler/abstractinterpretation.jl (hunk 4 above)
Comment (Keno) [reply base/compiler/abstractinterpretation.jl:1848]: This one I think in theory could be proven nothrow, even if we don't model the rt.

**Pull request merged**